Use a single-striped connection pool for each database layer instead of a single shared connection

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue that this PR relates to and which requirements it tackles. Jira issues of the form ADP- will be auto-linked. -->

ADP-586

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 73df15f60162b40b7dbd7166287d6154426f2312
  :round_pushpin: **use a single-striped connection pool for each database layer**

  It is a rather common practice to use a pool of database connection
  when dealing with databases. So far, we've been using a single shared
  connection per wallet worker with, in front of each connection a lock
  preventing concurrent access to the database. The lock is only
  necessary because of the way persistent handles query statements
  internally, in principle, SQLite handles concurrent database accesses
  just well.

  For basic wallets, this is a relatively useless change. But for larger
  wallets like those manipulated by exchanges, we've observed very slow
  response time due to concurrent access of the database lock. Indeed,
  some requests may grab the lock for 10 or 20 seconds, preventing any
  requests from going throug. However, most requests are read-only
  requests and could be executed in parallel, at the discretion of
  the SQLite engine. I hope that the introduction of a connection pool
  will improve the overall experience for large wallets by better
  serving concurrent requests on the database. Finger crossed.


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!--
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Jira will detect and link to this PR once created, but you can also link this PR in the description of the corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
 ✓ Finally, in the PR description delete any empty sections and all text commented in <!--, so that this text does not appear in merge commit messages.
-->

KtorZ committed:
use a single-striped connection pool for each database layer

  It is a rather common practice to use a pool of database connection
  when dealing with databases. So far, we've been using a single shared
  connection per wallet worker with, in front of each connection a lock
  preventing concurrent access to the database. The lock is only
  necessary because of the way persistent handles query statements
  internally, in principle, SQLite handles concurrent database accesses
  just well.

  For basic wallets, this is a relatively useless change. But for larger
  wallets like those manipulated by exchanges, we've observed very slow
  response time due to concurrent access of the database lock. Indeed,
  some requests may grab the lock for 10 or 20 seconds, preventing any
  requests from going throug. However, most requests are read-only
  requests and could be executed in parallel, at the discretion of
  the SQLite engine. I hope that the introduction of a connection pool
  will improve the overall experience for large wallets by better
  serving concurrent requests on the database. Finger crossed.

KtorZ commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -345,19 +360,38 @@ instance MatchMigrationError SqliteException where
+    let createConnection = do
+            let info = mkSqliteConnectionInfo connStr
+            conn <- Sqlite.open connStr
+            executeManualMigration migration conn
```
Note that I originally executed the migration outside of the `createConnection` with a bracket open / close / executeMigration ... and it was mostly fine except for database views in the DB.Pool. The database property tests would fail with a `no such table: <name of the view>` as if the migration had not been executed. 

I have looked a bit in the documentation but didn't find anything that suggested that views are scoped to a single connection but it seems to be the case. At least, it should be safe running migration on every connection anyway... but still strange.

rvl replied:
Hmm. It seems preferable for the migrations to be executed before other pool connections are allowed. Perhaps there needs to be a transaction commit or something?

rvl commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -219,32 +191,75 @@ startSqliteBackend
+    -- Note that `withResource` does already mask async exception but only for
+    -- dealing with the pool resource acquisition. The action is then ran
+    -- unmasked with the acquired resource. If an asynchronous exception occurs,
+    -- the resource is NOT placed back in the pool.
```
Docs say that an exception of _any type_ means that the resource is not placed back in the pool.
This will be quite helpful for error recovery I think.

KtorZ replied:
I hope.

rvl commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -345,19 +360,38 @@ instance MatchMigrationError SqliteException where
+        createConnection
+        destroyConnection
+        numberOfStripes
+        timeToLive
```
If the pool thread is cancelled (i.e. wallet is deleted or server needs to exit), those lingering pool connections are going to get cleaned up promptly, right?

KtorZ replied:
Yes. I swapped the `destroyDBLayer` function with:

```hs
destroyDBLayer :: Tracer IO DBLog -> SqliteContext -> IO ()
destroyDBLayer tr SqliteContext{connectionPool,dbFile} = do
    traceWith tr (MsgDestroyConnectionPool dbFile)
    destroyAllResources connectionPool
```

So this is done as part of the `withDBLayer` bracket. The same quirks as before with regards to closing a single SQLite connection still applies but `destroyConnection` now does exactly what `destroyDBLayer` was doing, that is, retry for some cherry-picked exceptions and handle already closed connections as well.

rvl commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -345,19 +360,38 @@ instance MatchMigrationError SqliteException where
+    timeToLive = 600 :: NominalDiffTime
+    -- When running in :memory:, we want a single connection that does not get
+    -- cleaned up.
+    maximumConnections = maybe 1 (const 10) fp
```
Is there some difference with sqlite in-memory databases which mean only one concurrent connection is possible? Otherwise, maximumConnections may as well be the same for both.

The number 10 seems fine for each wallet database. Perhaps in future, to prevent scaling bottlenecks, the maximum for the stake pools database could be scaled by the number of wallets.

KtorZ replied:
I originally wanted to use [createSqlitePool](https://hackage.haskell.org/package/persistent-sqlite-2.10.6.2/docs/Database-Persist-Sqlite.html#v:createSqlitePool) from `persistent` but it doesn't give any ways to access the underlying connection (only the persistent abstraction `SqlBackend`) which we do need in a few places for "hardcore error handling" and also custom migrations. 

The `createSqlitePool` function rightfully mentions:

> Note that this should not be used with the :memory: connection string, as the pool will regularly remove connections, destroying your database. Instead, use withSqliteConn.

I stole the default `10` max connections and `600` seconds time to live from the default chosen in persistent. Using `1` for in the in-memory database seems to work-around the issue nicely as resource-pool always keep at least one connection. Since this is only used in testing, I found it to be a simple and non-invasive resolution.

rvl reviewed (approved):
Great! This will help a lot. I don't know why we didn't do this earlier, since resource-pool is so easy to use. I actually assumed that the API server got to use a different DB connection from the restore thread - oops, wrong.

KtorZ committed:
Properly handle 'SQLITE_BUSY' in the context of a connection pool

  I ran into quite a few issues with the integration tests since the
  unliftio merge and rebase (I think, as I am pretty I did observe unit
  and integration tests doing just fine with the resource pool at least
  once). I've been investigating this for most of the day, and found a few
  interesting cases:

  (a) SQLite may return 'SQLITE_BUSY' on pretty much any requests if two
      concurrent write queries hit the engine; though we
      currently only catch this kind of exception when we try closing the
      database so I generalized a bit our error handling here.

  (b) It seems that calling destroyAllResources from resource-pool does not
      prevent new threads from acquiring new resources. And there's no way
      with the resource-pool library itself to prevent the creation of
      new resources after a certain point. So it may happen that while the
      database layer is being destroyed, new database connections are created
      and start causing conflicts between each others.

rvl commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -127,14 +137,26 @@ import qualified Database.Sqlite as Sqlite
-    { connectionPool :: Pool (SqlBackend, Sqlite.Connection)
+    { connectionPool :: Pool SqlBackend
     -- ^ A handle to the Persistent SQL backend.
+    , isDatabaseActive :: TVar Bool
```
A link to https://www.sqlite.org/wal.html#busy might help.

KtorZ committed:
use bracket-style resource acquisition for the db connection pool

  This avoids the need for an extra 'TVar Bool' to guard the connection
  pool from threads whishing to acquire new resources. Instead, we can
  wrap the pool acquisition in a bracket: `bracket createPool destroyAllResources`
  so that the pool is cleaned up when done and we are sure that no
  thread will attempt to acquire a new resource while
  destroyAllResources is called.

  This sole change wasn't as straightforward as I wanted because it
  moves the control of the `SqliteContext` up in the stack and therefore
  requires reviewing many more parts of both the pool and wallet db
  layers. I think it's for a greater good in the end and make them both
  slightly better / robust. In the end, it is still a bit "awkward" that
  we have constructors / functions in those modules that are solely used
  by the test code and not by the actual application (this is the case
  of 'withDBLayer' for instance...).

  To not over-complicate things, I ended up handling the in-memory and
  in-file SqliteContext setup a bit differently. Incidentally I realized
  later that we run most of our unit-tests on the 'in-memory' version;
  which means that we aren't testing the resource pool in the context of
  the unit tests. I am not sure whether this is a good thing or not: it
  makes the unit tests a bit more focus on testing the actual business
  logic, and we still have the system-level integration tests to put the
  resource pool under great stress.

rvl commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -167,9 +179,82 @@ queryLogFunc tr _loc _source level str = traceWith tr (MsgQuery msg sev)
+               -- runSqlConn is guarded with a lock because it's not threadsafe in
+               -- general.It is also masked, so that the SqlBackend state is not
+               -- corrupted if a thread gets cancelled while running a query.
+               -- See: https://github.com/yesodweb/persistent/issues/981
```
Perhaps in future we could drop the lock and mask_?
1. If we are confident that any pool connection will only be used by one thread at a time, then the lock isn't needed.
2. If the thread is cancelled during a database operation, then so be it. SQLite will rollback the uncommitted transaction. The connection won't be returned to the pool.

KtorZ replied:
(1) is only true if we can start the database in full_mutex I guess, which persistent doesn't really allow us to set that flag (although, we've forked it so it's easily added). This will not prevent retries however in case two writing threads are trying to write something at the same time, but at least it should give more concurrency to readers :thinking: Also "if we are confident that any pool connection will only be used by one thread at a time" ==> I have no such confidence at this stage unfortunately :( 

(2) Hmmm. That's right. Sounds scary though :scream:

rvl commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -202,49 +290,33 @@ destroyDBLayer (SqliteContext {getSqlBackend, trace, dbFile}) = do
+retryOnBusy :: Tracer IO DBLog -> IO a -> IO a
+retryOnBusy tr action =
+    recovering policy (mkRetryHandler isBusy) $ \RetryStatus{rsIterNumber} -> do
+        when (rsIterNumber > 0) $ traceWith tr (MsgRetryOnBusy rsIterNumber)
```
I like that there's logging. With the resource pool improvements. it will be interesting to see if this error ever happens in tests.

rvl reviewed (approved):
Excellent, thanks for this!

If I run the integration tests with `NO_CLEANUP=1`, then ctrl-c them, there are no longer `-shm` and `-wal` files left behind in the tests temporary directory. :tada:

Even better, wallets which were supposed to be deleted, actually seem to have been deleted.

KtorZ commented on lib/shelley/80-wallets-to-0.01.json:
```diff
@@ -0,0 +1 @@
+{"walletOverview":{"transactions":0,"addresses":20,"utxo":{"distribution":{"10000000000000000":0,"100000000000000":0,"1000000000000":0,"10000000000":0,"100000000":0,"1000000":0,"10000":0,"100":0,"1000000000":0,"100000000000":0,"10":0,"45000000000000000":0,"1000":0,"10000000000000":0,"1000000000000000":0,"100000":0,"10000000":0},"scale":"log10","total":{"quantity":0,"unit":"lovelace"}}},"listAddre ...
```
Whoops :point_up:

Anviking replied:
My bad, fixed.

rvl committed:
Regenerate nix

rvl committed:
newConnectionPool -> withConnectionPool

rvl committed:
No need for mask in runQuery

If the thread is cancelled, we want the query stopped immediately.

Anything uncommitted will be rolled back by persistent.

rvl committed:
Split Test.Utils.Resource.unBracket from Test.Hspec.Extra.aroundAll

This function will be useful in criterion benchmarks.

It also fixes handling of exceptions while allocating resources.

rvl committed:
Add tests for aroundAll

rvl committed:
Let bench:db work using withConnectionPool

rvl committed:
Straighten out connection pool logging a little

rvl committed:
Don't get stuck when deleting wallets

The deleteWallet handler opens a database connection then tries to
remove the database before closing the connection. But the sqlite
removeDatabase function waits for all connections to be closed first.

The livelock is only broken after the 1 minute timeout in
removeDatabase.

rvl committed:
Elide retryOnBusy logging

rvl committed:
Fix MsgApplyBlocks logging

It was bothering me.

rvl committed:
Simplify sqlite connection setup

KtorZ commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -423,13 +425,16 @@ instance ToText FollowLog where
+                [ "Applying block numbers [", buildRange blockHeights, "]"
+                , "  Wallet/node slots: ", slot (NE.last hdrs)
+                , "/", slot tipHdr
+                ]
```
:grimacing: Good.

rvl committed:
Switch checkpoint cache from IORef to MVar

rvl committed:
Straighten out SqliteContext

rvl committed:
Lightly refactor checkpoint cache

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1126,39 +1125,33 @@ newDBLayerWith cacheBehavior ti SqliteContext{runQuery} = do
+        alterCache :: W.Wallet s -> (Maybe (W.Wallet s) -> Maybe (W.Wallet s))
+        alterCache cp = \case
+            -- this seems suspicious
+            Just old | getHeight cp < getHeight old -> Just old
```
This is mainly there because of the state-machine tests. In the application, we do in practice never have the case where we try to write a checkpoint that is older than the latest checkpoint written because we process them in sequence from genesis to the tip, and we drop checkpoints upon rolling back. 

However, the state-machine tests aren't really respecting any of this and may insert checkpoints in any arbitrary order. That little guard is here to cope with this and enforce the invariant that the cache always contains the _most recent checkpoint_, not in terms of insertion order in the database but in terms of blockheight on the chain. 

I realize that this deserves a proper comment note...

rvl committed:
DBLayer: Attempt to make the checkpoint cache threadsafe

And add logging of database checkpoint cache

rvl committed:
Decouple DBFactory from DBLayer a little

rvl committed:
Add a query lock temporarily to let the checkpoint cache work

rvl committed:
Properly resolve merge conflicts from rebase
